src/utils/execution-order.ts

```diff
@@ -0,0 +1,147 @@
+import ExternalModule from '../ExternalModule';
+import Module from '../Module';
+import { randomUint8Array, Uint8ArrayEqual, Uint8ArrayXor } from './entryHashing';
+import error from './error';
+import { relative } from './path';
+import relativeId from './relativeId';
+
+interface OrderedExecutionUnit {
+	execIndex: number;
+}
+
+const compareExecIndex = <T extends OrderedExecutionUnit>(unitA: T, unitB: T) =>
+	unitA.execIndex > unitB.execIndex ? 1 : -1;
+
+export function sortByExecutionOrder(units: OrderedExecutionUnit[]) {
+	units.sort(compareExecIndex);
+}
+
+export function analyzeModuleExecution(
+	entryModules: Module[],
+	generateChunkHashes: boolean,
+	inlineDynamicImports: boolean,
+	manualChunkModules: Record<string, Module[]>
+) {
+	let curEntry: Module, curEntryHash: Uint8Array;
+	const cyclePaths: string[][] = [];
+	const allSeen: { [id: string]: boolean } = {};
+
+	const orderedModules: Module[] = [];
+	let analyzedModuleCount = 0;
+
+	const dynamicImports: Module[] = [];
+	const dynamicImportAliases: string[] = [];
+
+	let parents: { [id: string]: string };
+
+	const visit = (module: Module) => {
+		// Track entry point graph colouring by tracing all modules loaded by a given
+		// entry point and colouring those modules by the hash of its id. Colours are mixed as
+		// hash xors, providing the unique colouring of the graph into unique hash chunks.
+		// This is really all there is to automated chunking, the rest is chunk wiring.
+		if (generateChunkHashes) {
+			if (!curEntry.chunkAlias) {
+				Uint8ArrayXor(module.entryPointsHash, curEntryHash);
+			} else {
+				// manual chunks are indicated in this phase by having a chunk alias
+				// they are treated as a single colour in the colouring
+				// and aren't divisable by future colourings
+				module.chunkAlias = curEntry.chunkAlias;
+				module.entryPointsHash = curEntryHash;
+			}
+		}
+
+		for (const depModule of module.dependencies) {
+			if (depModule instanceof ExternalModule) {
+				depModule.execIndex = analyzedModuleCount++;
+				continue;
+			}
+
+			if (depModule.id in parents) {
+				if (!allSeen[depModule.id]) {
+					cyclePaths.push(getCyclePath(depModule.id, module.id, parents));
+				}
+				continue;
+			}
+
+			parents[depModule.id] = module.id;
+			if (!depModule.isEntryPoint && !depModule.chunkAlias) visit(depModule);
+		}
+
+		for (const dynamicModule of module.dynamicImportResolutions) {
+			if (!(dynamicModule.resolution instanceof Module)) continue;
```

Comment:
I guess dynamic imports to external modules are ok without an execIndex, although worth bearing this in mind from a data consistency point of view.

Reply:
Good point. Will change to make sure all (external and internal) modules are initialized consistently in case their `execIndex` is never set.